spec.html

```diff
@@ -1223,3 +1223,191 @@ <h1>Execution Contexts</h1>
+      <p>The source</p>
+      <pre><code class="javascript">
+        a = b + c
+        (d + e).print()
+      </code></pre>
+      <p>is <em>not</em> transformed by automatic semicolon insertion, because the parenthesized expression that begins the second line can be interpreted as an argument list for a function call:</p>
+      <pre><code class="javascript">a = b + c(d + e).print()</code></pre>
+      <p>In the circumstance that an assignment statement must begin with a left parenthesis, it is a good idea for the programmer to provide an explicit semicolon at the end of the preceding statement rather than to rely on automatic semicolon insertion.</p>
+    </emu-clause>
+
+    <emu-clause id="sec-hazards-of-automatic-semicolon-insertion">
+      <h1>Hazards of Automatic Semicolon Insertion</h1>
+      <em>This section is non-normative.</em>
+      <p>ECMAScript programs can be written in a style with very few semicolons, based on a heavy dependence on automatic semicolon insertion. However, as described above, semicolons are not inserted at every newline, and some of these cases which do not have an automatically inserted semicolon can be counter-intuitive or confusing.</p>
+
+      <p>As new syntactic features are added to ECMAScript, additional cases requiring explicit semicolons emerge over time. As such, consistently explicit semicolon use is recommended.</p>
+
+      <p>The term "automatic semicolon insertion hazard" refers, informally, to a place where a developer may expect a semicolon to be inserted, but according to the rules described above, it is not. The rest of this section describes a number of automatic semicolon insertion hazards in this version of ECMAScript.</p>
+      <emu-clause id="sec-asi-hazards-in-statement-lists">
+        <h1>Automatic Semicolon Insertion Hazards in Statement Lists</h1>
+        <p>In a |StatementList|, many |StatementListItem|s end in semicolons, which may be omitted using automatic semicolon insertion. As a consequence of the rules above, at the end of a line ending an expression, a semicolon will be required if the following line begins with any of the following:
+        <ul>
+          <li><strong>An opening parenthesis (<code>(</code>)</strong>. Without a semicolon, the two lines together will be treated as a function call.</li>
+          <li><strong>An opening square bracket (<code>[</code>)</strong>. Without a semicolon, the two lines together will be treated as property access, rather than an array literal.</li>
+          <li><strong>A template string (<code>`</code>)</strong>. Without a semicolon, the two lines together will be interpreted as tagged template literal, with the previous expression as the function tag.</li>
+          <li><strong>Unary <code>+</code> or <code>-</code></strong>. Without a semicolon, the two lines together will be interpreted as a usage of the corresponding infix operator.</li>
```

Comment:
If this list is intended to be complete, it's missing "regex literal":

```js
a
/b/i
```
[parses as](http://shift-ast.org/parser.html?type=Script&code=x%20=%20a%0a/b/i) two divisions, not an identifier followed by a regex literal.

Reply:
Good catch; added this. Do you think I should say that the list is exhaustive?

Reply:
I think it should specify clearly whether it is or isn't, and I'm pretty sure it now is and as such should say so. (Though it might be worth calling out that "unary minus" includes `-2` and similar.)

Reply:
One reason I didn't say it's exhaustive is because I don't have a perfectly formal definition of what an ASI hazard is. It's really resting on developer intuition. For example, one definition would be a situation with two lines of source text which are valid both if a semicolon is inserted and if not; however, this would exclude the generator method after initializer case (since it's just a SyntaxError), but I wanted to include this text to explain that it's necessary to write a semicolon there. OTOH we're not just talking about every place that there's a line continuation (e.g., putting something inside parentheses is definitely not an ASI hazard). Do you have another idea?

Reply:
A stab at a formal definition: Any place where removing a textually-present semicolon token at the end of a line, which is not parsed as an EmptyStatement, might change the semantics of the code (not counting F.p.toString). This list isn't exhaustive with respect to that definition - you'd have to include `for` statements - but it's close.